### PR TITLE
HAWQ-1003. Implement batched ACL check through Ranger.

### PR DESCRIPTION
**Pull request by zhangh43**
This pr includes :
HAWQ-1001. Implement HAWQ basic user ACL check through Ranger.
HAWQ-1002. Implement configuration GUC related to Ranger.
HAWQ-1003. Implement bached ACL check through Ranger.

when GUC enable_ranger is ON, we will use ranger_acl_check to determine where user has the permission to query the data.

### CODE REVIEW

**Comment by paul-guo-** [comment]
ORCA did ACL check in their code also so please make sure your batch code works for them also.

File: src/backend/parser/parse_relation.c [hunk 1]
```diff
@@ -2713,15 +2712,90 @@ warnAutoRange(ParseState *pstate, RangeVar *relation, int location)
...
+    ranger_check_args = lappend(ranger_check_args, ranger_check_arg);
+
+  } // foreach
+
+  // ranger ACL check with package Oids
+  List *aclresults = NIL;
+  aclresults = pg_rangercheck_batch(ranger_check_args);
+  if (aclresults == NIL)
+  {
+    elog(ERROR, "ERROR\n");
```
Comment (paul-guo-) [inline src/backend/parser/parse_relation.c]: I would think this log does not help debugging much. Could we provide more info?

File: src/include/utils/rangerrest.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*-------------------------------------------------------------------------
+ *
+ * rangerrest.h
+ *	routines to interact with Ranger REST API
```
Comment (paul-guo-) [inline src/include/utils/rangerrest.h]: REST API for Ranger interaction?

File: src/backend/utils/misc/guc.c [hunk 3]
```diff
@@ -6250,6 +6261,15 @@ static struct config_int ConfigureNamesInt[] =
...
 	{
+    {"hawq_rps_address_port", PGC_POSTMASTER, PRESET_OPTIONS,
+      gettext_noop("rps server address port number"),
+      NULL
+    },
+    &rps_addr_port,
+    1, 1, 65535, NULL, NULL
+  },
+
+	{
```
Comment (paul-guo-) [inline src/backend/utils/misc/guc.c]: What is rps? If this is Ranger specific, please specify in the descriptions.

File: src/backend/utils/misc/guc.c [hunk 4]
```diff
@@ -8154,6 +8174,15 @@ static struct config_string ConfigureNamesString[] =
...
 	{
+    {"hawq_rps_address_host", PGC_POSTMASTER, PRESET_OPTIONS,
+      gettext_noop("rps server address hostname"),
+      NULL
+    },
+    &rps_addr_host,
+    "localhost", NULL, NULL
+  },
+
+	{
```
Comment (paul-guo-) [inline src/backend/utils/misc/guc.c]: Same as before.

**Comment by paul-guo-** [comment]
Seems there are file conflicts with https://github.com/apache/incubator-hawq/pull/1046

At least rangerrest.h

File: src/backend/utils/misc/guc.c (hunk 3 above)
Comment (ictmalili) [inline src/backend/utils/misc/guc.c]: rps is short for Ranger Plugin Service, which stores HAWQ Ranger Plugin

File: src/backend/parser/parse_relation.c (hunk 1 above)
Comment (zhangh43) [inline src/backend/parser/parse_relation.c]: yes, this error message should be replaced by the exact error message in future. for example, which table doesn't pass the acl check.

File: src/include/utils/rangerrest.h (hunk 2 above)
Comment (zhangh43) [inline src/include/utils/rangerrest.h]: just using libcurl to connect to RPS(ranger plugin server). And RPS will provide REST API

File: src/backend/utils/misc/guc.c (hunk 4 above)
Comment (zhangh43) [inline src/backend/utils/misc/guc.c]: fixed

File: src/backend/utils/misc/guc.c (hunk 3 above)
Comment (zhangh43) [inline src/backend/utils/misc/guc.c]: fixed

**Pull request closed without merging**